tiatoolbox/tools/registration/wsi_registration.py

```diff
@@ -172,3 +176,259 @@ def match_histograms(image_a, image_b, kernel_size=7):
+    def extract_features(self, fixed_img, moving_img):
+        """CNN based feature extraction for registration.
+
+        This function extracts multiscale features from a pre-trained
+        VGG-16 model for an image pair.
+
+        Args:
+            fixed_img (:class:`numpy.ndarray`):
+                A fixed image.
+            moving_img (:class:`numpy.ndarray`):
+                A moving image.
+
+        Returns:
+            dict:
+                A dictionary containing the multiscale features.
+                The expected format is {layer_name: features}.
+
+        """
+        if len(fixed_img.shape) != 3 or len(moving_img.shape) != 3:
+            raise ValueError(
+                "The required shape for fixed and moving images is n x m x 3."
+            )
+
+        if fixed_img.shape[2] != 3 or moving_img.shape[2] != 3:
+            raise ValueError("The input images are expected to have 3 channels.")
+
+        self.xScale = 1.0 * np.array(fixed_img.shape[:2]) / self.patch_size
+        self.yScale = 1.0 * np.array(moving_img.shape[:2]) / self.patch_size
+        fixed_cnn = imresize(
+            fixed_img, output_size=self.patch_size, interpolation="linear"
+        )
+        moving_cnn = imresize(
+            moving_img, output_size=self.patch_size, interpolation="linear"
+        )
+
+        fixed_cnn = fixed_cnn / 255.0
+        moving_cnn = moving_cnn / 255.0
+
+        fixed_cnn = np.moveaxis(fixed_cnn, -1, 0)
+        moving_cnn = np.moveaxis(moving_cnn, -1, 0)
+
+        fixed_cnn = np.expand_dims(fixed_cnn, axis=0)
+        moving_cnn = np.expand_dims(moving_cnn, axis=0)
+        cnn_input = np.concatenate((fixed_cnn, moving_cnn), axis=0)
+
+        x = torch.from_numpy(cnn_input).type(torch.float32)
+        return self.FeatureExtractor(x)
+
+    @staticmethod
+    def finding_match(feature_dist):
+        """Computes matching points.
+
+        This function computes all the possible matching points
+        between fixed and moving images.
+
+        Args:
+            feature_dist (:class:`numpy.ndarray`):
+                A feature distance array.
+
+        Returns:
+            :class:`numpy.ndarray`:
+                An array of matching points.
+            :class:`numpy.ndarray`:
+                An array of floating numbers representing quality
+                of each matching points.
+
+        """
+        seq = np.arange(feature_dist.shape[0])
+        ind_first_min = np.argmin(feature_dist, axis=1)
+        first_min = feature_dist[seq, ind_first_min]
+        mask = np.zeros_like(feature_dist)
+        mask[seq, ind_first_min] = 1
+        masked = np.ma.masked_array(feature_dist, mask)
+        second_min = np.amin(masked, axis=1)
+        return np.array([seq, ind_first_min]).transpose(), np.array(
+            second_min / first_min
+        )
+
+    @staticmethod
+    def compute_feature_distance(feature_x, feature_y, factor):
```

Comment:
```suggestion
    def compute_feature_distances(features_x, features_y, factor):
```

```suggestion
    def compute_feature_distance_matrix(features_x, features_y, factor):
```

```suggestion
    @np.vectorize
    def compute_feature_distance(feature_x, feature_y, factor):
```

Suggest either renaming to make it clear that it is not computing a single distance, or make it into a [numpy vectorized function](https://numpy.org/doc/stable/reference/generated/numpy.vectorize.html) that does one or many.